### DISCUSSION TITLE
Proposal: Binary Distribution Format

### DISCUSSION THREAD

**Discussion by theory** [pgxn/discussion#2] [RFCs]
RFC: Binary Distribution Format
===============================

> ## NOTE: Moved to a [pull request](https://github.com/pgxn/rfcs/pull/2) for the [RFCs Book](http://pgxn.github.io/rfcs/).

*   **Tracking:** [RFC-2](https://github.com/pgxn/rfcs/pull/2)
*   **Author:** [theory](https://github.com/theory)
*   **Status:** **Pending**
*   **Task:** pgxn/planning#4

----

This RFC specifies the binary distribution format for PGXN packages, also called the trunk format.[^wheel]

A trunk is a ZIP-format archive with a specially formatted file name and the `.trunk` extension. It contains a single distribution nearly as it would be installed by [PGXS]. Although a specialized installer is recommended, a trunk file may be installed by simply copying directories of files to destinations defined by [pg_config].

# File Format

## File name convention

The trunk filename is:

```
{package}-{version}+{pg}-{platform}.trunk
```

Definition of variables:

*   `package`: Package name, e.g. `vector`, `postgis`, `pgAdmin`.
*   `version`: Distribution version in [SemVer] format without build metadata, e.g., `0.8.6` or `1.0.0-beta`.
*   `pg`: Major version of Postgres the binary was built against, e.g., `pg15`, `pg16`.
*   `platform`: The platform the binary was built for. Will be made up of one to three hyphen-delimited[^hyphen] values for the OS, version information[^PEPs], and CPU architecture. Examples: `any`, `gnulinux-amd64`, `darwin-23.5.0.arm64`, `musllinux-1.2-amd64v3`. The allowed values will be defined one or more separate RFCs.

### Examples:

*   `pgtap-1.0.1+pg15-any.trunk` packages `pgtap` version 1.9.1, compatible with Postgres 15 (any minor release) on any platform.

*   `pair-0.32.1+pg16-gnulinux-amd64.trunk` packages `pair` version 0.32.1,
    compatible with Postgres 16 (any minor release) on GNU libc-based Linux for amd64 CPUs.

*   `pair-0.32.1+pg16-darwin-23.5.0-arm64.trunk` packages `pair` version 0.32.1, compatible with Postgres 16 (any minor release) on Darwin version 23.5.0 (macOS) for arm64 CPUs.

### Escaping and Unicode

The `+` in the file name indicates the division between the package name and version and the package metadata. The package name and version must not include a `+`. This allows the file name, without the `.trunk` extension, to also function as a valid [SemVer].

Tools producing trunks should verify that the filename components do not contain `+`, as the resulting file may not be processed correctly if it does.

The package name should be lowercase.

The file name components should all be UTF-8.

The filenames *inside* the archive are encoded as UTF-8. Although some ZIP clients in common use do not properly display UTF-8 filenames, the encoding is supported by the ZIP specification.

### Parsing

Parsing of the file name takes place in four parts:

1.  For the file name, remove the `.trunk` extension. If working with the directory name (prefix) extracted from the archive, there will be no `.trunk` extension.

2.  Split the name into two parts at the `+` sign. The left part is the package name and [SemVer]. The right part is the platform specification.

3.  For the left part, split on the right-most dash. If the string to the right of the dash is a valid [SemVer], then the left part is the package name. If the right string is not a valid [SemVer], try again at the second right-most dash and check again. Continue until a valid SemVer is produced or else fail.

4.  Split the right string on dashes. There will be between two and four values as follows:

    *   Two: the postgres version (`pg16`) and `any`.
    *   Three: the postgres version (`pg16`), the OS (`gnulinux`, `darwin`, etc.), and the architecture (`amd64`, `arm64`, etc.)
    *   Four: the postgres version (`pg16`), the OS (`gnulinux`, `darwin`, etc.), the OS version (`23.5.0`) and the architecture (`amd64`, `arm64`, etc.)

#### Examples:

*   `pgtap-1.0.1+pg15-any`
    *   Package: `pgtap`
    *   Version: `1.0.1`
    *   Postgres: `pg15`
    *   Platform: `any`
*   `pair-0.32.1-beta1+pg16-gnulinux-amd64`
    *   Package: `pair`
    *   Version: `0.32.1-beta1`
    *   Postgres: `pg16`
    *   OS: `gnulinux`
    *   Architecture: `amd64`
*   `pair-0.32.1+pg16-darwin-23.5.0-arm64`
    *   Package: `pair`
    *   Version: `0.32.1`
    *   Postgres: `pg16`
    *   OS: `darin`
    *   OS Version: `23.5.0`
    *   Architecture: `arm64`

## File contents

The contents of a wheel file should unpack into a directory with the same name as the file, but without the `.trunk` extension. The contents of the directory are:

*   `trunk.json` contains metadata necessary to install the extension. The format wil be subject to a future RFC, but at a minimum will include the trunk format version, package version, dependencies, license, language and runtime (e.g., libc implementation and version), platform metadata, and Postgres version and build configuration. Trunk installers should warn if the trunk version is greater than the version it supports, and must fail if the Trunk version has a greater major version than the version it supports.
*   `digests` contains a list of (almost) all the files in the trunk and their secure hashes. Each line lists a single file and its checksum in the [BSD digest format]: `{algorithm} ({filename}) = {checksum}`. Every file except `digests` --- which cannot contain a hash of itself --- must be listed in this file. The cryptographic hash algorithm must be [SHA-256] or better; specifically, MD5 and SHA-1 are not permitted, as signed trunk files rely on the strong hashes in `digests` to validate the integrity of the archive.
*   The `pgsql` directory contains one or more subdirectories named for `pg_config` directory configurations: `bin`, `doc`, `html`, `include`, `pkginclude`, `lib`, `pkglib`, `locale`, `man`, `share`, and `sysconf`. Each contains the files to be installed in the corresponding `pg_config` directory.
*   Dynamic language scripts must appear in `pgsql/bin` and begin with exactly `'#!{cmd}`, where `cmd` is the name of the interpreter, in order to enjoy script wrapper generation and shebang rewriting at install time. They must have no extension. The list of supported interpreters will depend on the features of the installer, but one can reasonably expect support for [Perl], [Python], and [Ruby]. If no appropriate instance of the given interpreter is present, the installer may abort the installation.
*   `README`, `LICENSE`, and `CHANGELOG` may optionally be in the directory. Each must be plain text or Markdown-formatted. In the latter case, they may use the extension `.md`.
*   Trunk, being an installation format intended to install pre-compiled binaries and supporting files, does not include a `Makefile`, `configure` file or any other tool for building the package contents.

During extraction, trunk installers verify all the hashes in `digests` against the file contents. Apart from `digests` and its signatures, installation will fail if any file in the archive is not both mentioned and correctly hashed in `digests`.

# Details

## Installing a Trunk

The following descriptions will use a trunk file named `pair-0.32.1+pg16-gnulinux-amd64.trunk`. Trunk installation notionally consists of two phases:

1.  Unpack
    *   Validate digests. Ensure that every file, aside from `digests` itself, is listed in `digest` along with it valid hash digest. If any file is missing or has an invalid digest, installation should fail. If a file listed in `digests` is not present, installation should fail.
    *   Parse the `trunk.json` file. Check that the distribution is compatible with:
        *   The trunk format version
        *   The platform (OS, OS version, and architecture); `any` is allowed for any platform
        *   The PostgreSQL version
2.  Install
    *   If applicable, update scripts starting with `#!{cmd}` to point to the correct interpreter. Fail if no such interpreter is present.
    *   Iterate over each subdirectory of the `pgsql` directory.
    *   If the directory corresponds to a directory configuration from [pg_config], install its contents in that target directory.

  [^wheel]: With much inspiration and from and gratitude to the [Python Binary distribution format].
  [^hyphen]: Why hyphens? They allow the entire file name, between the package name and the `.trunk` extension, to be a valid [SemVer].
  [^PEPs]: See for example [PEP 600] defining Python wheel tags for different versions of GNU libc and [PEP 656] defining tags for different versions of musl libc. See also how [Homebrew] uses [macOS version names] in file names for its packages.

# References

*   [Python Binary distribution format]
*   [trunk POC]

  [PGXS]: https://www.postgresql.org/docs/current/extend-pgxs.html
    "PostgreSQL Docs: Extension Building Infrastructure"
  [pg_config]: https://www.postgresql.org/docs/current/app-pgconfig.html
    "PostgreSQL Docs: pg_config"
  [Python Binary distribution format]: https://packaging.python.org/en/latest/specifications/binary-distribution-format/
  [SemVer]: https://semver.org "Semantic Versioning 2.0.0"
  [PEP 600]: https://peps.python.org/pep-0600/
    "PEP 600 – Future ‘manylinux’ Platform Tags for Portable Linux Built Distributions"
  [PEP 656]: https://peps.python.org/pep-0656/
    "PEP 656 – Platform Tag for Linux Distributions Using Musl"
  [Homebrew]: https://brew.sh "Homebrew: The Missing Package Manager for macOS (or Linux)"
  [macOS version names]: https://github.com/oras-project/oras/issues/237#issuecomment-815250008
    "oras-project/oras#237 Comment from sjackman"
  [BSD digest format]: https://stackoverflow.com/q/1299833/79202
  [SHA-256]: https://en.wikipedia.org/wiki/SHA-2 "Wikipedia: SHA-2"
  [Perl]: https://perl.org "The Perl Programming Language"
  [Python]: https://python.org "The Python Programming Language"
  [Ruby]: https://ruby-lang.org/en/ "The Ruby Programming Language"
  [trunk POC]: https://gist.github.com/theory/7dc164e5772cae652d838a1c508972ae
    "trunk POC using PGXS, bash, tar, shasum, and jq"

**Comment by theory** [comment]
# Other Ideas

Some other ideas for the format, in either the short or long term:

*   Adopt the [Python wheel signing pattern]
*   Include an [SPDX SBOM](https://spdx.dev)?
*   Allow non-postgres libraries to be included, such as OS dependencies, either in the appropriate `pgsql` subdirectory or perhaps in a separate `sys` directory

  [Python wheel signing pattern]: https://packaging.python.org/en/latest/specifications/binary-distribution-format/#signed-wheel-files
    "Python Binary distribution format: Signed wheel files"

**Reply by MMeent** [reply]
That link does not lead to wheel signing patterns but to a Postgres error on Stack Overflow; could you provide a working link?

**Reply by theory** [reply]
Oops, paste fail. Fixed.

**Comment by einhverfr** [comment]
I think you also need to take into account different glibc versions and different standard C library implementations.  So you probably also need a libc implementation field and a libc version number field also.

In some distros one can even choose between MUSL and GlibC on many source-based distros for example.  And there are binary MUSL-based distros also.

**Reply by theory** [reply]
That makes a lot of sense. This RFC doesn't cover the metadata for the `trunk.json` file, but I'll add these to the short list of examples here. Thanks for the feedback!

**Reply by einhverfr** [reply]
It's not just metadata though.  If I want to ship a library that builds against MUSL libc and one that builds against glibc, both on Linux, we would have a naming collision.  Similarly if I build against one version of glibc, and run against an earlier one I will get errors.....  So the version is also potentially important.

**Reply by theory** [reply]
You mean a naming collision in the file name? I'm thinking the metadata built into the file name should be relatively limited.

How does one work with community yum and apt packages today when upgrading the OS? If you upgrade from Debian bullseye to bookworm (presumably with different versions of libc), do you have to update all the apt packages, too?

**Reply by einhverfr** [reply]
So if I want to build for Bookwork and Alpine, how does that work?  Both are linux so it seems like the file names should be the same

**Reply by einhverfr** [reply]
It is worth noting that Python's wheel format ended up having to add both these fields (a libc and a libc version) to avoid these sorts of problems.

**Reply by theory** [reply]
Ah, perhaps the distribution name should be used instead of just "linux"?

I don't see libc mentioned in the [Python wheel spec](https://packaging.python.org/en/latest/specifications/binary-distribution-format/); the format is:

```
{distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-{platform tag}.whl.
```

For the platform tag, the example is `linux_x86_64`, which is equivalent to `linux-amd64` in this proposal. But presumably they document the allowed values elsewhere. Which is why I think we can be flexible here at the start and define additional values and make adjustments as we try things out.

**Reply by einhverfr** [reply]
See PEP 600  and the discussion at https://github.com/pypa/manylinux
Also see https://peps.python.org/pep-0656/

**Reply by theory** [reply]
Oh, interesting. Looks like `manylinux` means "glibc linux". Odd name. I guess it's intended to mean, Debain, Ubuntu, RedHat, and other glibc-based distributions.

As for the version, perhaps it makes sense to revise the `os_version` field to be less specific. Then it could be used for the libc version for gnu linux, the musl version for musl linux, and the OS version for darwin.

**Reply by theory** [reply]
Okay, I've collapsed the os, os_version, and arch fields into a single notional field, "platform", which contains the same basic parts, but what the supported values will be will be defined later. I added a footnote referencing the use of gnu libc and musl libc versions, as well as Homebrew macOS versions, to get the point across, and updated examples to use `gnulinux` to get the point across.

**Comment by MMeent** [comment]
> arch: The CPU architecture the binary was build for, e.g., amd64, arm64. Required except when os is "any". The allowed values are defined by the source for [go tool dist list](https://github.com/golang/go/blob/release-branch.go1.22/src/cmd/dist/build.go#L1682-L1741).

I suspect that restriction makes this very restrictive for higher-performance versions of extensions that depend on modern CPUs, like pgvector: The list contains only archetypes (`amd64`), and not the more modern ISA levels like `x86-64-v3` that guarantee support for some set of instructions, such as the POPCNT instruction that's only guaranteed for `x86-64-v3` and up, or AVX/AVX2 which are only in `x86-64-v3` and up.

**Reply by theory** [reply]
That's fair, but I think that information can be part of additional metadata in the `trunk.json` file (and also the OCI manifest; see [this POC](https://justatheory.com/2024/06/trunk-oci-poc/ "POC: Distributing Trunk Binaries via OCI")). The idea is that the 90% use case is covered by this high-level platform info, and an installing client that needs to target a more specific processor can find all the options and select the best one based on its metadata.

**Reply by MMeent** [reply]
Well, that's also part of the concern. 
If we don't differentiate it in the trunk filename, then we won't have unique filenames for trunks for the same CPU archetype but different ages of development, right?

At some point a difference in features becomes a new kind of cpu altogether, even if they're backward compatible with other compiled binaries; similar to x86 vs x86-64: technically you can run x86 on x86-64, but usually you prefer a binary built specifically for 64 bits.

So, how would one distribute a binary for x86-64-v1 for general distribution, and one for x84-64-v4 for modern x86-64 CPUs (Intel archs after 2017; AMD Zen 4+)?

**Reply by theory** [reply]
We could include such info in the filename, yes, although with OCI distribution it doesn't matter (the file name becomes irrelevant). But I'm trying to follow precedents here. Perhaps I need to go back and review what RPM and Yum do, though I think it's mainly that different architectures are in different directories…

**Reply by theory** [reply]
It looks like yum.postgresql.org doesn't use anything other than the [basic architectures](https://yum.postgresql.org/packages/#pg17). I see:

* x86_64 (aka amd64)
* aarch64 (aka arm64)
* ppc64le
* i386

Looking through the [list of repos for RedHat 12](https://download.postgresql.org/pub/repos/yum/12/redhat/) I don't see anything different. AFAIC they don't support any specific release of x86.

That said, we can always expand the necessary annotations later, but I definitely would like to include them in `trunk.json` from the start.

**Reply by MMeent** [reply]
There is [a discussion ongoing on -hackers](https://www.postgresql.org/message-id/flat/CA%2BhUKGKS64zJezV9y9mPcB-J0i%2BfLGiv3FAdwSH_3SCaVdrjyQ%40mail.gmail.com) about approximately this topic (updating the baseline x86-64 architecture), which is why I decided to call the same issue out over here, too.

**Reply by theory** [reply]
Ah, useful thread, thanks. I'll happily follow whatever the community Linux registries standardize on. And if that means adding stuff like `x86_64v3` or whatever, then hey, no problem.

**Reply by theory** [reply]
The second part of the change I [mentioned upthread(#discussioncomment-9885707) is that it no longer specifies that the list of OSes and architectures will come from `go tool`, but from future RFCs, and I included an example, `amd64v3` to suggest that there may be more architecture tags to come.

**Comment by MMeent** [comment]
> one can reasonably expect support for [Perl](https://perl.org/), [Python](https://python.org/), and [Ruby](https://ruby-lang.org/en/).

I'm not so sure about these requirements. I don't have Python on my system, nor Ruby, and Perl is only installed because PostgreSQL requires it for development. 

Would Python/Perl/Ruby be an installation requirement for Trunk to work with this? That sounds like a rather large set of dependencies.

**Reply by theory** [reply]
No. My plan is to build SDKs and a CLI [in Rust](https://justatheory.com/2024/05/pgxn-language-poll-result/). The only other tooling required to build a trunk package would be whatever an extension requires (PGXS, compiler, pgrx, etc.).

**Reply by theory** [reply]
Oh, sorry, what I mean is: if an extension requires a Perl, Python, or Ruby pipeline, the PGXN trunk CLI should be able to use it to build the extension and the trunk file.

**Reply by MMeent** [reply]
The exerpt was from

> Dynamic language scripts must appear in pgsql/bin and begin with exactly '#!{cmd}, where cmd is the name of the interpreter, in order to enjoy script wrapper generation and shebang rewriting at install time. They must have no extension. The list of supported interpreters will depend on the features of the installer, but one can reasonably expect support for [Perl](https://perl.org/), [Python](https://python.org/), and [Ruby](https://ruby-lang.org/en/).

Which implies that the trunk file's contents may include scripts later executed on the client systems, using the client's Perl/Python/Ruby - and that's where my issue lies. 

Is that a misinterpretation on my part?

**Reply by theory** [reply]
You are not. I've included this because a fair number of extensions ship that use these interpreters. dalibo/emaj, for example, installs both [Perl and PHP scripts](/dalibo/emaj/tree/v4.4.0/client). I want to support that pattern. Ideally the installer CLI will either prompt the user to install the required interpreters, or else install them directly by shelling out to `yum` or whatever.

The same issue applies for extensions that require third-party libraries, BTW, like PostGIS. Third-party dependencies are gonna be a PITA to get right, but we gotta get em right.

**Reply by theory** [reply]
I've added this sentence to the end of that section:

> If no appropriate instance of the given interpreter is present, the installer may abort the installation.

**Comment by pramsey** [comment]
For extensions that use external dependencies (simples ones like pgsql-http depending on libcurl, complex ones like postgis depending on geos/gdal/proj/etc) what is the expectation for how those dependencies should be bundled? Static linking only? Is there some way to place dynamic dependencies in the pgsql tree such that the deps in the tree are picked up preferentially over deps already on the system?

**Reply by theory** [reply]
I've had this in the back of my mind for a while; thanks for asking about it.

This RFC makes no assumptions about external dependencies, but hopefully it leaves room for it. Before long I want to do some experiments to determine patterns; pgsql-http will be a pretty ideal place to start, since there's just the one dependency to work with. 

How I suspect it will come down, though, is that there will be a number of possibilities:

1.  Statically compile the dependency. Simplest solution for packagers, of course, as long as the licensing is clear (IIUC, Yum wants a list of the licenses of all dependencies).

2.  Include the library and compile it as a separate DSO and install it along with your main DSO. This is the pattern I need to try, but I suspect it will work, and perhaps will be the default solution. Unless multiple extension use the same library and might stomp on each other. In which case, see option 3.

3.  Create a separate extension distribution that only includes the library (libcurl in this case) and installs it where Postgres wants DSOs. Then identify that package as a dependency. This would allow multiple extensions to depend on the same package without stomping on each other, and would, I think, work today with PGXN dependency resolution.

4.  Identify external dependencies in some sort of platform-neutral manner and then teach the installer how to resolve them. I suspect we'll need something like this for more complicated projects like PostGIS and PostgresML, but I don't know how well it will work. I included some notes about it in the [dependencies section of the PGXN Metadata Sketch](https://justatheory.com/2024/03/rfc-pgxn-metadata-sketch/#dependencies), and included this note:

    > The external field is the tricky one. How does one provide a canonical dependency for a third-party library or app that may have loads of different names on different platforms, different packaging systems (apt vs rpm), and even different versions of them, or split apart for development and runtime use? Some possibilities:
    >
    > * Use [purl](https://github.com/package-url/purl-spec) and let downstream services handle the conversion
    > * Create a rule set that allows one to predict dependency name changes across ecosystems, similar to [repology-rules](https://github.com/repology/repology-rules)
    > * Just use [Repology](https://repology.org/)?

I haven't had a chance to follow up yet, but as I've started [drafting v2 of the metadata spec](https://github.com/pgxn/pgxn-meta-spec/pull/5), I expect to spend some time with all of these ideas soon.

**Reply by df7cb** [reply]
Or just acknowledge that .deb and .rpm have solved the "library dependencies" problem for 3 decades, and not try to reinvent the wheel in an incomplete manner. Sorry, I don't have a less sarcastic answer here.

**Reply by theory** [reply]
No worries, @df7cb. That's kind of what I'm trying to get at with option 4: use the OS packaging system, if available, to try to satisfy dependencies.

Nice thing about the first three options, though, is that one doesn't have to depend on or wait for the packaging system to build the packages — especially since not all extensions are available in all the packaging systems.

Anyway, we have _options_.

**Reply by pramsey** [reply]
There's a reason pip, cpan, cran, gem, etc, exist and include facilities for shipping and building deps or just flat out shipping binaries. There's really no way to build a packaged postgis "trunk" that depends on system libraries and says nothing except "I'm a postgis trunk for linux ix86". Because that doesn't even begin to describe all the variability in what might be on the system (or, more likely, not on the system). Basically the "trunk" system is only useful for people who have a perfectly calibrated set of system packages, which is to say that people might as well not use "trunk" and should just be using the... existing RPM and DEB packages from PGDG. Like, why are we here at all?

**Reply by theory** [reply]
Most of my experience of systems like this is CPAN, and there's quite a lot of variation in how external dependencies are handled. Some include the binaries (or sources), some rely on the system to provide dependencies (quite common IME), some require an "Alien" CPAN module to provide the dependencies, and some do all three and let the person installing the module choose. I suspect it's much the same for the other systems.

I'm a little confused by this statement:

> Basically the "trunk" system is only useful for people who have a perfectly calibrated set of system packages

Because the first three options I mentioned above are about embedding compiled dependencies directly in a trunk distribution, or depending on other trunk distributions that include needed binaries. This is not unlike RPM, DEB, and other packaging models.

> should just be using the... existing RPM and DEB packages from PGDG. Like, why are we here at all?

Not every extension is available via DEB and RPM. Sometimes they get behind. And not everyone uses Linux! So if you prefer to use DEB and RPM, that's great, I see no eroding of those systems.

But there are so many more extensions and OSes to support, and I'd like to build a Postgres-based (which is to say, `pg_config`-based) system that can be generalized for use in many places.

**Reply by pramsey** [reply]
Sorry I spoke inelegantly. Basically my thrust is that "just use system packaging" is antithetical to what seems (to me) to be the point of the "trunk" story.  Once one accepts system packaging as the answer to dependencies, one is on a fast and straight road that leads (IMO) logically to "only use system packaging". Which isn't what this conversation is about.

**Reply by theory** [reply]
Ah, thank you. Yes, that makes much more sense. And it's why I'm thinking much more, from the start, about the first three options, and expect to defer the fourth until it's needed for some reason — if ever.

But I'd like to see how far I can get with pgsql-http, then PostGIS to prove those patterns. Should be interesting!

**Comment by theory** [comment]
NOTE: Moved to a [pull request](https://github.com/pgxn/rfcs/pull/2) for the [RFCs Book](http://pgxn.github.io/rfcs/).
